src/Microsoft.Extensions.Http/DefaultHttpClientFactory.cs

```diff
@@ -66,9 +67,20 @@ private HttpMessageHandler CreateHandler(string name)
             var builder = _services.GetRequiredService<HttpMessageHandlerBuilder>();
 
             var options = _optionsMonitor.Get(name);
-            for (var i = 0; i < options.HandlerBuilderActions.Count; i++)
+
+            // Group each policy into ordered stages ordered by the 'Stages' ordered by the position
+            // in the 'stages' listing. Skip over anything without a defined order. 
+            var stages = options.HttpMessageHandlerPolicies
+                .GroupBy(policy => policy.Stage)
+                .OrderBy(group => options.HttpMessageHandlerPolicies.Stages.IndexOf(group.Key))
+                .Where(group => options.HttpMessageHandlerPolicies.Stages.Contains(group.Key));
+
+            foreach (var stage in stages)
             {
-                options.HandlerBuilderActions[i](builder);
+                foreach (var policy in stage)
+                {
+                    policy.Apply(builder);
+                }
```

Comment:
This is how the stages work. 

The fact that we ignore policies without a matching stage is a feature not a bug. You can easily take out things you don't want to run this way. We don't expect adding a new stage to users do often.